internal/gcpmanager/main.go

```diff
@@ -0,0 +1,109 @@
+// Package gcp provides a GCP implementation of the Substratus Cloud Interface (SCI).
+package gcpmanager
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	credentials "cloud.google.com/go/iam/credentials/apiv1"
+	storage "cloud.google.com/go/storage"
+	sci "github.com/substratusai/substratus/internal/sci"
+	credentialspb "google.golang.org/genproto/googleapis/iam/credentials/v1"
+)
+
+// server implements the sci.ControllerServer interface.
+type Server struct {
+	sci.UnimplementedControllerServer
+	StorageClient *storage.Client
+}
+
+// CreateSignedURL generates a signed URL for a specified GCS bucket and object path.
+func (s *Server) CreateSignedURL(ctx context.Context, req *sci.CreateSignedURLRequest) (*sci.CreateSignedURLResponse, error) {
+	bucketName, objectName := req.GetBucketName(), req.GetObjectName()
+
+	bucket := s.StorageClient.Bucket(bucketName)
+	obj := bucket.Object(objectName)
+	if _, err := obj.Attrs(ctx); err != nil {
+		if err != storage.ErrObjectNotExist {
+			// An error occurred that was NOT ErrObjectNotExist.
+			// This is an unexpected error and we should return it.
+			return nil, err
+		}
+		// the object doesn't exist and we want to continue and create the signed URL.
+	}
+
+	saEmail, err := getServiceAccountEmail()
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := credentials.NewIamCredentialsClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &storage.SignedURLOptions{
+		// this is optional storage.SigningSchemeV2 is the default. storage.SigningSchemeV4 is also available.
+		Scheme: storage.SigningSchemeV2,
+		Method: http.MethodPost,
+		Headers: []string{
+			"Content-Type:multipart/form-data",
+		},
+		Expires:        time.Now().Add(time.Duration(req.GetExpirationSeconds()) * time.Second),
+		GoogleAccessID: saEmail,
+		SignBytes: func(b []byte) ([]byte, error) {
+			req := &credentialspb.SignBlobRequest{
+				Payload: b,
+				Name:    saEmail,
+			}
+			resp, err := c.SignBlob(ctx, req)
+			if err != nil {
+				panic(err)
+			}
+			return resp.SignedBlob, err
+		},
+	}
+
+	// Create a signed URL
+	url, err := storage.SignedURL(bucketName, objectName, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sci.CreateSignedURLResponse{Url: url}, nil
+}
+
+func getServiceAccountEmail() (string, error) {
```

Comment:
potentially a simpler path here: https://pkg.go.dev/golang.org/x/oauth2/google#FindDefaultCredentials

Reply:
Most google clients can be initialized like:

```go
		compute, err := computev1.NewService(context.Background())
		if err != nil {
```

which will find default credentials automatically

Reply:
Which should work out of the box with workload identity... also you should be able to make it work with some gclouding locally.

Reply:
Maybe I haven't dug into this deeply enough but I wonder if the compute client actually gives what I'm after here... Is there any example you have of querying the metadata service but falling back on static creds?

I really just need the SA email address of the credentail being used. Maybe the simplest way to do that is to invoke different functions based on the presence of a static credential (SA key).

Reply:
FIgured this out with the metadata package. thanks for the pointer there 🙏